packages/context/src/context.ts

```diff
@@ -34,16 +43,90 @@ export class Context {
   protected _parent?: Context;
 
   /**
-   * Create a new context
+   * A list of registered context observers
+   */
+  protected readonly observers: Set<ContextObserver> = new Set();
+
+  /**
+   * Internal counter for pending events which observers have not processed yet
+   */
+  private pendingEvents = 0;
+
+  /**
+   * Create a new context. For example,
+   * ```ts
+   * // Create a new root context, let the framework to create a unique name
+   * const rootCtx = new Context();
+   *
+   * // Create a new child context inheriting bindings from `rootCtx`
+   * const childCtx = new Context(rootCtx);
+   *
+   * // Create another root context called "application"
+   * const appCtx = new Context('application');
+   *
+   * // Create a new child context called "request" and inheriting bindings
+   * // from `appCtx`
+   * const reqCtx = new Context(appCtx, 'request');
+   * ```
    * @param _parent The optional parent context
+   * @param name Name of the context, if not provided, a `uuid` will be
+   * generated as the name
    */
   constructor(_parent?: Context | string, name?: string) {
+    super();
     if (typeof _parent === 'string') {
       name = _parent;
       _parent = undefined;
     }
     this._parent = _parent;
     this.name = name || uuidv1();
+
+    this.setupEventHandlers();
+  }
+
+  /**
+   * Set up an internal listener to notify registered observers asynchronously
+   * upon `bind` and `unbind` events
+   */
+  private setupEventHandlers() {
+    // Ideally p-event should allow multiple event types in an iterator
+    this.observeEvent('bind');
```

Comment:
This function is observing multiple event instances, I think `observeEvents` or `observeEventInstances` would be a more appropriate name.

`observeEvent` is returning a promise that you are ignoring here. I find this difficult to spot and possibly misleading. At minimum, please add a comment to explain that `observeEvent` is an async function that we are going to leave running in background. Can we come up with a name that will convey this information without the need for an additional comment?

Reply:
fixed